// markdown-vscode-contributions/dist/cjs/enums.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Contributions = void 0;
let Contributions;
exports.Contributions = Contributions;
(function (Contributions) {
  Contributions["AUTHENTICATION"] = "authentication";
  Contributions["BREAKPOINTS"] = "breakpoints";
  Contributions["CODEACTIONS"] = "codeActions";
  Contributions["COLORS"] = "colors";
  Contributions["COMMANDS"] = "commands";
  Contributions["CONFIGURATION"] = "configuration";
  Contributions["CONFIGURATIONDEFAULTS"] = "configurationDefaults";
  Contributions["CUSTOMEDITORS"] = "customEditors";
  Contributions["CONTINUEEDITSESSION"] = "continueEditSession";
  Contributions["CSSCUSTOMDATA"] = "css.customData";
  Contributions["DEBUGGERS"] = "debuggers";
  Contributions["DOCUMENTATION"] = "documentation";
  Contributions["GRAMMARS"] = "grammars";
  Contributions["HTMLCUSTOMDATA"] = "html.customData";
  Contributions["HTMLLANGUAGEPARTICIPANTS"] = "htmlLanguageParticipants";
  Contributions["ICONS"] = "icons";
  Contributions["ICONTHEMES"] = "iconThemes";
  Contributions["JSONVALIDATION"] = "jsonValidation";
  Contributions["KEYBINDINGS"] = "keybindings";
  Contributions["LANGUAGES"] = "languages";
  Contributions["LOCALIZATIONS"] = "localizations";
  Contributions["MARKDOWNMARKDOWNITPLUGINS"] = "markdown.markdownItPlugins";
  Contributions["MARKDOWNPREVIEWSCRIPTS"] = "markdown.previewScripts";
  Contributions["MARKDOWNPREVIEWSTYLES"] = "markdown.previewStyles";
  Contributions["MENUS"] = "menus";
  Contributions["NOTEBOOKPRELOAD"] = "notebookPreload";
  Contributions["NOTEBOOKRENDERER"] = "notebookRenderer";
  Contributions["NOTEBOOKS"] = "notebooks";
  Contributions["PROBLEMMATCHERS"] = "problemMatchers";
  Contributions["PROBLEMPATTERNS"] = "problemPatterns";
  Contributions["PRODUCTICONTHEMES"] = "productIconThemes";
  Contributions["REMOTEHELP"] = "remoteHelp";
  Contributions["RESOURCELABELFORMATTERS"] = "resourceLabelFormatters";
  Contributions["SEMANTICTOKENMODIFIERS"] = "semanticTokenModifiers";
  Contributions["SEMANTICTOKENSCOPES"] = "semanticTokenScopes";
  Contributions["SEMANTICTOKENTYPES"] = "semanticTokenTypes";
  Contributions["SNIPPETS"] = "snippets";
  Contributions["SUBMENUS"] = "submenus";
  Contributions["TASKDEFINITIONS"] = "taskDefinitions";
  Contributions["TERMINALPROFILES"] = "terminal";
  Contributions["THEMES"] = "themes";
  Contributions["TYPESCRIPTSERVERPLUGINS"] = "typescriptServerPlugins";
  Contributions["VIEWS"] = "views";
  Contributions["VIEWSCONTAINERS"] = "viewsContainers";
  Contributions["VIEWSWELCOME"] = "viewsWelcome";
  Contributions["WALKTHROUGHS"] = "walkthroughs";
})(Contributions || (exports.Contributions = Contributions = {}));
